### ISSUE TITLE
Mobile application

### ISSUE THREAD

**Issue by Chocobozzz** [Chocobozzz/PeerTube#250]
Now we documented our REST API (/!\ still in Alpha /!\) maybe some people will want to create a mobile application. Here is an issue to discuss about this.

Just  remember that PeerTube uses WebTorrent and thus WebRTC to seed videos. If you want to keep compatibility with WebRTC, you'll need to develop the application in JavaScript and put it in a container that support WebRTC (like Electron on Desktop).

If you don't want to keep the compatibility with WebRTC, you'll loose web browser seeders. But you could leech/seed with TCP/UDP instead so it could be fine too.

**Comment by rigelk** [comment]
There are WebRTC implementations in [Google Android](https://webrtc.org/native-code/android/) and [Apple iOS](https://webrtc.org/native-code/ios/) too.

**Comment by sylflo** [comment]
I could be interested to the Android app. The problem is that even tough I have knowledge on Android, I have not so much clue on our WebRTC works. I followed some tutorials about this technology but I have to admit not understand it very well.

I could make a protoype first or if someone already started something I would be happy to join

**Comment by Shotman** [comment]
Couldn't you use Cordova or Ionic to create the mobile app since you use angular ?

**Comment by rigelk** [comment]
As some people [pointed out on the forum](https://framacolibri.org/t/peertube-sur-android/2260) [in french], if we target mobile platforms it might also be interesting to extend to TV boxes (Android TV especially).

Also, NewPipe already has [though a lot about the user interace](https://github.com/TeamNewPipe/NewPipe/issues/650) and [aims at supporting additional platforms](https://github.com/TeamNewPipe/NewPipe/pull/693). Maybe we could look for a synergy here.

**Comment by sschueller** [comment]
I believe the only way to go is writing a native app (at least on the Android side). Yes it is a lot more work but if Peertube is wants to be a competing platform to youtube the mobile experience needs to be as close as possible,

Android does not have a full javascript engine built in like iOS so apps made in Cordova, Ionic Appcelerator are fairly large and don't produce a good user experience.

I like the NewPipe idea but I don't see it being a replacement for a native Peertube app. Remember NewPipe can not be put in the play store because it violates google/youtube's rules.

**Comment by rigelk** [comment]
All good points @sschueller. 

Nothing prevents us from forking NewPipe and actually replace access to YouTube with PeerTube. Even though the experience won't be as close as a native app of our own, it will still allow us to bootstrap faster.

In the meantime, I still believe we should try to make the web interface as good as possible a PWA: no design shift, no language shift, no need to separate forces, and no need for end users to install another app.

---
@sylflo are you still interested making a native app?

**Comment by sylflo** [comment]
@rigelk yes of course but I am too busy these days probably until the end of April.
For me if we only talk about user experience, the best thing like you said is to have a pwa first so the users will not need to install an application. I know it's something which works pretty well since most big companies did it so their users stay longer on their app/websites.

I would really like to participate in the project but right now is not possible for me.

**Comment by sschueller** [comment]
I started work on an Android version however it is still very very early stages: https://github.com/sschueller/peertube-android . Please let me know if you have any feedback, comments etc.

Thanks.

**Comment by Nutomic** [comment]
FYI NewPipe has support for different sites now, so it would be possible to integrate PeerTube without forking it.

**Comment by Nutomic** [comment]
https://github.com/TeamNewPipe/NewPipeExtractor/issues/79

**Comment by strypey** [comment]
Working with any existing video app/ plug-in to support PeerTube as well as YT (or whatever they current work with) would be a great way to bring more developer interest to the PT project, and avoid fragmenting the energy of the PT team across a number of apps. I agree with @rigelk that priority #1 is making the web app UX awesome on every device with a web browser, but it's great to see folks taking the initiative to start working on a native app ([ideally not using Electron](https://sircmpwn.github.io/2016/11/24/Electron-considered-harmful.html)). 

One thing I've been thinking about when it comes to apps is how to do branding in ways that Jo User finds less confusing. If I was running a PT instance, I would make sure it had a distinctive look and feel (while still giving due credit to PT), and if there are client-side PT apps, I would consider forking them to apply the same UX design. That way people can use StrypeyTube (or whatever) in their browser, and download the StrypeyTube app, but obviously I would make it very clear in both UI and all documentation what the app was forked from, and that it can be used to watch videos on any PR site.

**Comment by mcb2003** [comment]
One other, perhaps less thought about positive of integrating  peertube into the NewPipe app,  or even building a native app is screen reader accessibility for the Blind and visually impaired. I’m almost completely blind and find  NewPipe  very accessible with the talkback screen reader on android. Native apps also work much better than WebApps, although I must say the main peertube web  Interface actually works remarkably well considering I doubt many people have thought about screenreader accessibility yet.

**Comment by maddanio** [comment]
If I may: what would also help is improve mobile experience. On iOS for example it’s possible to make a web site feel quite native, I.e. give it an icon on the home screen and no Navbar and all. Maybe even some caching and stuff. Similar things are happening on android and I think There is also standardization attempts. If you want I can do some research. Unfortunately I am not well set up for web hacking myself, though I hope I can get round to changing that sometime :)
